bpf/tc_http_tp.c

```diff
@@ -0,0 +1,588 @@
+
+    index &= MAX_IP_PACKET_SIZE;
+
+    const struct datasum_loop_ctx *lctx = (const struct datasum_loop_ctx *)ctx;
+
+    const unsigned char *b = lctx->b;
+    const unsigned char *e = lctx->e;
+    __u32 *sum = lctx->sum;
+
+    if (!b) {
+        return 1;
+    }
+
+    if (&b[index] + sizeof(__u16) > e) {
+        return 1;
+    }
+
+    __u16 word;
+
+    __builtin_memcpy(&word, &b[index], sizeof(word));
+
+    *sum += word;
+
+    return 0;
+}
+
+static __always_inline void
+calculate_datasum(const unsigned char *b, const unsigned char *e, __u32 *sum) {
+    if (b >= e) {
+        *sum = 0;
+        return;
+    }
+
+    const __u32 len = (e - b) & MAX_IP_PACKET_SIZE;
+
+    struct datasum_loop_ctx datasum_loop_ctx = {b, e, sum};
+
+    bpf_loop(len, calculate_datasum_loop, &datasum_loop_ctx, 0);
+
+    if (len % 2 == 1) {
+        if (&b[len - 1] >= e) {
+            *sum = 0;
+            return;
+        }
+
+        __u16 word = 0;
+        __builtin_memcpy(&word, &b[len - 1], 1);
+
+        *sum += word;
+    }
+
+    while (*sum >> 16) {
+        *sum = (*sum & 0xFFFF) + (*sum >> 16);
+    }
+}
+
+static __always_inline void
+calculate_datasum_inline(const unsigned char *b, const unsigned char *e, __u32 *sum) {
+    if (b >= e) {
+        *sum = 0;
+        return;
+    }
+
+    const __u32 len = (e - b) & MAX_INLINE_LEN;
+
+    for (__u32 i = 0; i < len; i += 2) {
+        if (&b[i] + 1 >= e) {
+            break;
+        }
+
+        __u16 word;
+
+        __builtin_memcpy(&word, &b[i], 2);
+
+        *sum += word;
+    }
+
+    if (len % 2 == 1) {
+        if (&b[len - 1] >= e) {
+            *sum = 0;
+            return;
+        }
+
+        __u16 word = 0;
+        __builtin_memcpy(&word, &b[len - 1], 1);
+
+        *sum += word;
+    }
+
+    while (*sum >> 16) {
+        *sum = (*sum & 0xFFFF) + (*sum >> 16);
+    }
+}
+
+static __always_inline void
+encode_hex_skb(unsigned char *dst, const unsigned char *src, __u32 src_len) {
+
+#pragma clang loop unroll(full)
+    for (__u32 i = 0, j = 0; i < src_len; i++) {
+        unsigned char p = src[i];
+
+        dst[j++] = hex[(p >> 4) & 0xff];
+        dst[j++] = hex[p & 0x0f];
+    }
+}
+
+// this "beauty" ensures we hold pkt in the same register being range
+// validated
+static __always_inline unsigned char *
+check_pkt_access(unsigned char *buf, __u32 offset, const unsigned char *end) {
+    unsigned char *ret;
+
+    asm goto("r4 = %[buf]\n"
+             "r4 += %[offset]\n"
+             "if r4 > %[end] goto %l[error]\n"
+             "%[ret] = %[buf]"
+             : [ret] "=r"(ret)
+             : [buf] "r"(buf), [end] "r"(end), [offset] "i"(offset)
+             : "r4"
+             : error);
+
+    return ret;
+error:
+    return NULL;
+}
+
+static __always_inline void
+make_tp_string_skb(unsigned char *buf, const tp_info_t *tp, const unsigned char *end) {
+    buf = check_pkt_access(buf, EXTEND_SIZE, end);
+
+    if (!buf)
+        return;
+
+    *buf++ = 'T';
```

Comment:
This keeps the verifier happy, both in terms of register offset validation, and in terms of keeping the program complexity at bay.